torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -0,0 +1,222 @@
+    // same scalar type as input tensors.
+    st = typesFromTensors[0];
+    if (std::any_of(typesFromTensors.begin(), typesFromTensors.end(), [&st](const c10::ScalarType& type) {
+      return type != st;
+    })) {
+      std::cerr << "Warning: ONNX Scalar Type Analysis - Scalar types mismatch for tensor inputs of operator "
+                << n->kind().toDisplayString()
+                << ". Please report a bug to PyTorch. "
+                << "The scalar type "
+                << c10::toString(*st)
+                << " of the first tensor is chosen." << std::endl;
+    }
+  } else {
+    // When inputs consist of only scalars.
+    st = PromoteScalarTypes(typesFromScalars);
+  }
+
+  return st;
+}
+
+static void UpdateScalarTypeForInputs(Node* n, const c10::ScalarType& scalar_type) {
+  const int64_t onnx_type = ScalarTypeToONNXType(scalar_type);
+  if (onnx_type < 0) {
+    std::cerr << "Warning: ONNX Scalar Type Analysis - Scalar type: "
+              << c10::toString(scalar_type)
+              << " of input tensor in operator: " << n->kind().toDisplayString()
+              << " not supported in ONNX. " << std::endl;
+    return;
+  }
+
+  for (auto input : n->inputs()) {
+    auto input_tensor_type = input->type()->cast<TensorType>();
+    auto input_scalar_type = input_tensor_type->scalarType();
+
+    if ((input->node()->kind() == onnx::Constant) ||
+        (input_scalar_type && (*input_scalar_type != scalar_type))) {
+      if (input->node()->kind() == onnx::Constant) {
+        // Fix up the scalar directly instead of inserting a cast operator.
+        // NOTE: Keep only the else branch once constant_folding is enabled by
+        // default.
+        at::Tensor val = input->node()->t(attr::value);
+        at::Tensor new_val = val.to(scalar_type);
+        Node* const_node = n->owningGraph()->create(onnx::Constant);
+        const_node->t_(attr::value, new_val);
+        const_node->insertBefore(n);
+        const_node->output()->setType(TensorType::create(new_val));
+        n->replaceInputWith(input, const_node->output());
+      } else {
+        Node* cast_node = n->owningGraph()->create(onnx::Cast);
+        cast_node->addInput(input);
+        cast_node->i_(attr::to, onnx_type);
+        cast_node->insertBefore(n);
+        cast_node->output()->setType(CreateProfiledTensorTypeWithScalarType(
+            input_tensor_type, scalar_type));
+        n->replaceInputWith(input, cast_node->output());
+      }
+    }
+  }
+}
+
+static void UpdateScalarTypeForOutput(Node* n, const c10::ScalarType& scalar_type) {
+  auto output_tensor_type = n->output()->type()->cast<TensorType>();
+  n->output()->setType(
+      CreateProfiledTensorTypeWithScalarType(output_tensor_type, scalar_type));
+}
+
+static void ImplicitCastForONNX(Block* block) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
+    for (auto sub : it->blocks()) {
+      ImplicitCastForONNX(sub);
+    }
+    WithInsertPoint guard(*it);
```

Comment:
Looks like you are only using `insertBefore` for new nodes insertion. That function ignores the graph insert point, so I think you could delete this `WithInsertPoint guard`.

Reply:
Thanks!